README.md
This library provides classes to encode/decode bencoded bytes.
Mainly used to implement bittorrent related functions.


## Features

- Decode/encode bencoded bytes.

## Usage

Example:

```dart
import 'dart:typed_data';
import 'package:bencode/bencode.dart';

main() async {
  var decoded = bDecoder.convertString('d5:hello5:worlde');
  // or final decoded = bDecoder.convert('d5:hello5:worlde'.runes.toString());
  print(decoded);
  // Output: {hello: world}
  print(decoded is Map);
  // Output: true

  decoded = bDecoder.convertString('i123e');
  print(decoded);
  // Output: 123
  print(decoded is int);
  // Output: true

  decoded = bDecoder.convertString('li123ei123ee');
  print(decoded);
  // Output: [123, 123]
  print(decoded is List);
  // Output: true

  decoded = bDecoder.convertString('11:hello world');
  print(decoded);
  // bencoded strings are converted to bytes by default.
  // Output: [104, 101, 108, 108, 111, 32, 119, 111, 114, 108, 100]
  print(String.fromCharCodes(decoded));
  // Output: hello world
  print(decoded is List<int>);
  // Output: true

  var encoded = bEncoder.convert({'hello': 'world'});
  print(encoded);
  // Output: [100, 53, 58, 104, 101, 108, 108, 111, 53, 58, 119, 111, 114, 108, 100, 101]
  print(String.fromCharCodes(encoded));
  // Output: d5:hello5:worlde

  encoded = bEncoder.convert(123);
  print(String.fromCharCodes(encoded));
  // Output: i123e

  encoded = bEncoder.convert([123, 456]);
  print(String.fromCharCodes(encoded));
  // Output: li123ei456ee

  encoded = bEncoder.convert(Uint8List.fromList([123, 123]));
  print(encoded);
  // Output: [50, 58, 123, 123]
  print(String.fromCharCodes(encoded));
  // Output: 2:{{
}

```

Decode bittorrent files:
```dart
import 'dart:io';
import 'package:bencode/bencode.dart';

main() async {
  final data = File('test/multi.torrent').readAsBytesSync();
  final decoded = bDecoder.convert(data);
  print(decoded.keys);
  // Output: (announce, announce-list, created by, creation date, encoding, info)
}
```

## Tests

Run tests:
```
pub run test
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: http://github.com/xtyxtyx/bencode/issues


CHANGELOG.md
## 0.1.0+1

- Update examples


## 0.1.0

- Initial version, created by Stagehand


test/bencode_test.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:torrent_bencode/torrent_bencode.dart';
import 'package:test/test.dart';

void main() {
  group('bDecoder tests', () {
    test('Convert torrent file', () async {
      final data = await File('test/multi.torrent').readAsBytes();
      expect(bDecoder.convert(data), anything);
    });

    test('Convert int', () {
      expect(bDecoder.convertString('i123e'), equals(123));
      expect(bDecoder.convertString('i456e'), equals(456));
      expect(bDecoder.convertString('i-123e'), equals(-123));
    });

    test('Convert dict', () {
      expect(bDecoder.convertString('d3:agei123e4:name4:xutye'),
          equals({'name': 'xuty'.runes.toList(), 'age': 123}));
    });

    test('Convert list', () {
      expect(
          bDecoder.convertString('l3:agei123e4:name4:xutye'),
          equals([
            'age'.runes.toList(),
            123,
            'name'.runes.toList(),
            'xuty'.runes.toList(),
          ]));
    });

    test('Convert string', () {
      expect(bDecoder.convertString('11:hello world'),
          equals('hello world'.runes.toList()));
      expect(bDecoder.convertString('4:hola'), equals('hola'.runes.toList()));
      expect(() => bDecoder.convertString('100:haha'),
          throwsA(equals('broken string at 4')));
    });
  });

  group('bEncoder tests', () {
    test('Encode string', () {
      expect(bEncoder.convert('hello world'),
          equals('11:hello world'.runes.toList()));
    });

    test('Encode cjk', () {
      expect(bEncoder.convert('你好，世界'), equals(utf8.encode('15:你好，世界')));
    });

    test('Encode integer', () {
      expect(bEncoder.convert(123), equals('i123e'.runes.toList()));
    });

    test('Encode list', () {
      expect(bEncoder.convert([123, 'xuty']),
          equals('li123e4:xutye'.runes.toList()));
    });

    test('Encode bytes', () {
      expect(bEncoder.convert(Uint8List.fromList([0xAB, 0xCD])),
          equals('2:\u00ab\u00cd'.runes.toList()));
    });

    test('Encode map', () {
      expect(bEncoder.convert({'name': 'xuty'}),
          equals('d4:name4:xutye'.runes.toList()));
    });

    test('decode and re-encode torrent file', () async {
      final data = await File('test/multi.torrent').readAsBytes();
      final decoded = bDecoder.convert(data);
      final encoded = bEncoder.convert(decoded);
      expect(encoded, equals(data));
    });
  });
}


test/data_reader_test.dart
import 'dart:convert';
import 'package:torrent_bencode/src/data_reader.dart';
import 'package:test/test.dart';

void main() {
  group('DataReader tests', () {
    DataReader reader;

    setUp(() {
      final data = 'hello world'.runes.toList();
      reader = DataReader(data);
    });

    test('takeString', () {
      expect(reader.peekString(5), equals('hello'));
      expect(reader.peekString(5), equals('hello'));
      expect(reader.peekString(1), equals('h'));
      expect(reader.takeString(5), equals('hello'));
      expect(reader.takeString(), equals(' '));
      expect(reader.takeString(5), equals('world'));
      expect(reader.takeString(), equals(null));
      expect(reader.takeString(5), equals(null));
    });

    test('match', () {
      expect(reader.match('hello'), equals('hello'));
      expect(reader.match('hello'), equals('hello'));
      expect(reader.match('hola'), equals(null));
    });

    test('matches', () {
      expect(reader.matches('hello'), equals(true));
      expect(reader.matches('hello'), equals(true));
      expect(reader.matches('hola'), equals(false));
    });

    test('readInt', () {
      final reader = DataReader.fromString(('123abc456def-789'));
      expect(reader.readInt(), equals(123));
      expect(reader.expect('abc'), anything);
      expect(reader.readInt(), equals(456));
      expect(reader.expect('def'), anything);
      expect(reader.readInt(), equals(-789));
      expect(reader.atEOF, equals(true));
    });

    test('atEOf', () {
      expect(reader.atEOF, equals(false));
      expect(reader.takeString(11), equals('hello world'));
      expect(reader.atEOF, equals(true));
    });

    test('expect', () {
      expect(
          () => reader.expect('hola'), throwsA(equals('hola expected at 0')));
      expect(reader.expect('hello'), equals('hello'));
      expect(reader.expect(' '), equals(' '));
      expect(reader.expect('world'), equals('world'));
      expect(() => reader.expect('sekai'),
          throwsA(equals('sekai expected at 11')));
    });
  });
}


lib/src/bdecoder.dart
import 'dart:convert';

import 'data_reader.dart';

final bDecoder = BDecoder._();

class BDecoder extends Converter<List<int>, dynamic> {
  DataReader _reader;

  BDecoder._();

  /// `convert` converts bencoded string into `String`,
  /// bencoded dictionary into `Map<String, dynamic>`,
  /// bencoded list into `List<dynamic>`,
  /// and bencoded integers into `int`.
  dynamic convert(List<int> bytes) {
    _reader = DataReader(bytes);
    return _parse();
  }

  /// convertString is equivalent to `convert`
  /// except that convertString takes a String as its parameter.
  dynamic convertString(String str) {
    return convert(str.runes.toList());
  }

  /// tryConvert is equivalent to `convert`
  /// except that tryConvert wiil return null on failure rather than throw an exception
  dynamic tryConvert(List<int> bytes) {
    try {
      return convert(bytes);
    } catch (_) {
      return null;
    }
  }

  dynamic _parse() {
    _reader.reset();
    return _readNext();
  }

  dynamic _readNext() {
    if (_reader.matches('d')) {
      return _readDict();
    }
    if (_reader.matches('i')) {
      return _readInt();
    }
    if (_reader.matches('l')) {
      return _readList();
    }
    if (_reader.matches(RegExp(r'[0-9]'))) {
      return _readString();
    }
    if (_reader.atEOF) {
      return null;
    }

    throw 'd, i, l or <number> expected at ${_reader.pos}';
  }

  List<int> _readString() {
    final len = _reader.readInt();
    _reader.expect(':');
    final str = _reader.take(len);
    if (str == null) {
      throw 'broken string at ${_reader.pos}';
    }
    return str;
  }

  List<dynamic> _readList() {
    final result = <dynamic>[];
    _reader.expect('l');
    while (!_reader.matches('e')) {
      final value = _readNext();
      result.add(value);
    }
    _reader.expect('e');
    return result;
  }

  Map<String, dynamic> _readDict() {
    final result = <String, dynamic>{};
    _reader.expect('d');
    while (!_reader.matches('e')) {
      final key = utf8.decode(_readString());
      final value = _readNext();
      result[key] = value;
    }
    _reader.expect('e');
    return result;
  }

  int _readInt() {
    _reader.expect('i');
    final result = _reader.readInt();
    _reader.expect('e');
    return result;
  }
}


lib/src/bencoder.dart
import 'dart:convert';
import 'dart:io';

import 'dart:typed_data';

final bEncoder = BEncoder._();

class BEncoder extends Converter<dynamic, List<int>> {
  BEncoder._();

  /// `convert` converts dart object to bencoded bytes
  /// 
  /// Note that Strings are encoded in utf8. To encode raw bytes,
  /// please use Uint8List. More details in README and bencode_example.dart.
  List<int> convert(dynamic object) {
    if (object is String) {
      return _encodeUtf8String(object);
    }
    if (object is int) {
      return _encodeInteger(object);
    }
    if (object is Uint8List) {
      return _encodeString(object);
    }
    if (object is List) {
      return _encodeList(object);
    }
    if (object is Map) {
      return _encodeDictionary(object);
    }

    throw 'unsupported object type ${object.runtimeType}';
  }

  List<int> _encodeUtf8String(String str) {
    final builder = BytesBuilder();
    final data = utf8.encode(str);
    final length = ascii.encode(data.length.toString());
    const comma = 58; // ascii code of `:`
    builder.add(length);
    builder.addByte(comma);
    builder.add(data);
    return builder.toBytes();
  }

  List<int> _encodeString(List<int> data) {
    final builder = BytesBuilder();
    final length = ascii.encode(data.length.toString());
    const comma = 58; // ascii code of `:`
    builder.add(length);
    builder.addByte(comma);
    builder.add(data);
    return builder.toBytes();
  }

  List<int> _encodeInteger(int i) {
    return ascii.encode('i${i.toString()}e');
  }

  List<int> _encodeList(List<dynamic> list) {
    final builder = BytesBuilder();
    const i = 108; // ascii code of `l`
    const e = 101; // ascii code of `e`
    builder.addByte(i);
    for (var item in list) {
      builder.add(convert(item));
    }
    builder.addByte(e);
    return builder.toBytes();
  }

  List<int> _encodeDictionary(Map<String, dynamic> dict) {
    final builder = BytesBuilder();
    const d = 100; // ascii code of `d`
    const e = 101; // ascii code of `e`
    builder.addByte(d);
    for (var item in dict.entries) {
      builder.add(convert(item.key));
      builder.add(convert(item.value));
    }
    builder.addByte(e);
    return builder.toBytes();
  }
}


example/bencode_example.dart
import 'dart:typed_data';
import 'package:torrent_bencode/torrent_bencode.dart';

main() async {
  var decoded = bDecoder.convertString('d5:hello5:worlde');
  // or final decoded = bDecoder.convert('d5:hello5:worlde'.runes.toString());
  print(decoded);
  // Output: {hello: world}
  print(decoded is Map);
  // Output: true

  decoded = bDecoder.convertString('i123e');
  print(decoded);
  // Output: 123
  print(decoded is int);
  // Output: true

  decoded = bDecoder.convertString('li123ei123ee');
  print(decoded);
  // Output: [123, 123]
  print(decoded is List);
  // Output: true

  decoded = bDecoder.convertString('11:hello world');
  print(decoded);
  // bencoded strings are converted to bytes by default.
  // Output: [104, 101, 108, 108, 111, 32, 119, 111, 114, 108, 100]
  print(String.fromCharCodes(decoded));
  // Output: hello world
  print(decoded is List<int>);
  // Output: true

  var encoded = bEncoder.convert({'hello': 'world'});
  print(encoded);
  // Output: [100, 53, 58, 104, 101, 108, 108, 111, 53, 58, 119, 111, 114, 108, 100, 101]
  print(String.fromCharCodes(encoded));
  // Output: d5:hello5:worlde

  encoded = bEncoder.convert(123);
  print(String.fromCharCodes(encoded));
  // Output: i123e

  encoded = bEncoder.convert([123, 456]);
  print(String.fromCharCodes(encoded));
  // Output: li123ei456ee

  encoded = bEncoder.convert(Uint8List.fromList([123, 123]));
  print(encoded);
  // Output: [50, 58, 123, 123]
  print(String.fromCharCodes(encoded));
  // Output: 2:{{
}
